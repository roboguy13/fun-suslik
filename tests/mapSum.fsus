%generate sum [Sll[readonly]] Int
%generate mapSum [ListOfListsLayout[readonly]] Sll

data List := Nil | Cons Int List;

data ListOfLists := LNil | LCons List ListOfLists;

Sll : List >-> layout[x];
Sll (Nil) := emp;
Sll (Cons head tail) := x :-> head, (x+1) :-> tail, Sll tail;

ListOfListsLayout : ListOfLists >-> layout[x];
ListOfListsLayout (LNil) := emp;
ListOfListsLayout (LCons head tail) :=
  x :-> head, (x+1) :-> tail, ListOfListsLayout tail,
  Sll head;

sum : (Ptr List) -> Int;
sum (Nil) := 0;
sum (Cons head tail) := head + (sum tail);

mapSum : ListOfLists -> List;
mapSum (LNil) := Nil;
mapSum (LCons xs xss) :=
  Cons (instantiate [Sll[readonly]] Int sum (addr xs))
       (mapSum xss);

