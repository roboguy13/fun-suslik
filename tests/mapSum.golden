// *** Layout predicates ***

predicate ro_Sll(loc x) {
| (x == 0) => { emp }
| (not (x == 0)) => { [x,2] ** x :=> head ** (x+1) :=> tail ** ro_Sll(tail) }
}
predicate ro_ListOfListsLayout(loc x) {
| (x == 0) => { emp }
| (not (x == 0)) => { [x,2] ** x :=> head ** (x+1) :=> tail ** ro_ListOfListsLayout(tail) ** ro_Sll(head) }
}


// *** Function predicates ***

predicate mapSum__ro_Sll__ro_ListOfListsLayout(loc __p_x0, loc __r_x) {
| (__p_x0 == 0) => { __r_x == 0 ; emp }
| (not (__p_x0 == 0)) => { __r_x == __p_1 ; __p_x0 :-> xs ** (__p_x0+1) :-> xss ** ro_Sll(xs) ** [__p_x0,2] ** func sum__Int__ro_Sll(xs, __p_1) ** mapSum__ro_Sll__ro_ListOfListsLayout(xss, __p_x4) ** (__r_x+1) :-> __p_x4 ** [__r_x,2] }
}


// *** Function specifications ***


void mapSum(loc x1, loc r)
  { ro_ListOfListsLayout(x1) ** r :-> initialVal }
  { mapSum__ro_Sll__ro_ListOfListsLayout(x1, r0) ** r :-> r0 }
{ ?? }

{r :-> initialVal ** ro_ListOfListsLayout(x1)<_alpha_5>}
{r :-> r0 ** mapSum__ro_Sll__ro_ListOfListsLayout(x1, r0)<_alpha_6>}
void mapSum (loc x1, loc r) {
  if (x1 == 0) {
    *r = 0;
  } else {
    let h = *x1;
    let t = *(x1 + 1);
    mapSum(t, r);
    let r01 = *r;
    let r0 = malloc(2);
    *r = r0;
    *(r0 + 1) = r01;
    sum__Int__ro_Sll(h, r0);
  }
}


--- Source AST size: 97

--- SuSLik AST size: 153
Succeeded
